From 4d29882cda522ba40d3e8ac63f0ca2d325737ce3 Mon Sep 17 00:00:00 2001
From: Dmitry Ivanov <idmitry@qti.qualcomm.com>
Date: Wed, 14 May 2025 18:22:06 +0200
Subject: [PATCH] Add rgba resources backed by EGL image as not supporting
 view.

It was observed on sa8155 platform RGBA buffers which are backed by
EGL Image imported from GBM allocated buffers susceptible to swizzling.

Signed-off-by: Dmitry Ivanov <idmitry@qti.qualcomm.com>
---
 src/vrend_renderer.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/src/vrend_renderer.c b/src/vrend_renderer.c
index 041a03e1..73614eb1 100644
--- a/src/vrend_renderer.c
+++ b/src/vrend_renderer.c
@@ -997,6 +997,11 @@ bool vrend_format_is_bgra(enum virgl_formats format) {
            format == VIRGL_FORMAT_B8G8R8A8_SRGB);
 }
 
+static bool vrend_format_is_rgba(enum virgl_formats format) {
+   return (format == VIRGL_FORMAT_R8G8B8A8_UNORM ||
+           format == VIRGL_FORMAT_R8G8B8X8_UNORM);
+}
+
 static GLuint vrend_resource_get_internal_format_override(const struct vrend_resource *res)
 {
    /* Some shared resources imported to guest mesa as EGL images occupy 24bpp instead of more common 32bpp.
@@ -1031,7 +1036,7 @@ static bool vrend_resource_supports_view(const struct vrend_resource *res,
     * For views that don't require colorspace conversion, we can add swizzles
     * instead. For views that do require colorspace conversion, manual srgb
     * decode/encode is required. */
-   return !(vrend_format_is_bgra(res->base.format) &&
+   return !((vrend_format_is_bgra(res->base.format) || vrend_format_is_rgba(res->base.format)) &&
             has_bit(res->storage_bits, VREND_STORAGE_EGL_IMAGE)) &&
          (vrend_resource_get_internal_format_override(res) == GL_NONE);
 }
-- 
2.25.1

